### DISCUSSION TITLE
Use virtual consoles to launch a command e.g. xfreerdp

### DISCUSSION THREAD

**Discussion by icevk77** [ltsp/ltsp#514] [Q&A]
I am trying to get a functionality that older ltsp releases had. 
For example a **screen script** that launches **xfreerdp** on virtual console 8 (ALT-CTRL-F8) and was called from old config file lts.conf like this:
SCREEN_08="xfreerdp -f --plugin rdpsnd server"

I am testing it in ltsp server itself before moving it to the ltsp image.
What I have created so far is a systemd service **/etc/systemd/system/ltsp-vconsole.service**
```
[Unit]
Before=network.target
[Service]
Type=oneshot
ExecStart=openvt -f -c8 -- su -c 'xinit  /usr/bin/vconsole-xfreerdp -display :8 -- :8 vt8'
RemainAfterExit=yes
[Install]
WantedBy=multi-user.target
```
I enabled it
`systemctl enable lsa-vconsole.service`

I created the file **/usr/bin/vconsole-xfreerdp** as follows
```
#!/bin/bash
while [ "1" ]; do
        xfreerdp /f  /bpp:16  /cert-ignore -sec-nla /v:192.168.1.15
 done
```
`chmod +x /usr/bin/vconsole-xfreerdp`

and reboot.

This worked on ltsp server itself. 
Using the chrootless image installation it worked also on the client

So far it worked.

I am willing to write down documentation on https://github.com/ltsp/ltsp/wiki but I don't know if this is the right way to do it. 
 
Any help or guidance will be appreciated.

**Comment by alkisg** [comment]
The old LTSP was disabling inittab and was managing screens by itself.
Unfortunately, systemd, wayland, and changes in display managers made that old method unsuitable.

I don't think anyone in LTSP has researched "how to do this properly in the systemd/wayland era".
Also, personally I don't have enough free time currently to research this.

The good news is that it's not LTSP specific, so if you google around and make it work in a non-LTSP installation, it'll also work in LTSP installations.

**Comment by icevk77** [comment]
This is what I have done so far. It is not optimal but works.

Create a directory to store vconsole scripts
``mkdir /usr/share/ltsp/client/custom``

Create startup service file **/etc/systemd/system/ltsp-vconsole.service**
```
[Unit]
DefaultDependencies=no
Before=lightdm.service network.target halt.target shutdown.target reboot.target
[Service]
Type=simple
ExecStart=/usr/share/ltsp/client/custom/vconsole start
RemainAfterExit=yes
[Install]
WantedBy=multi-user.target halt.target shutdown.target reboot.target
```

Create common function and variable file **/usr/share/ltsp/client/custom/common**
```
#!/bin/bash
customscriptdir=/usr/share/ltsp/client/custom
function keep_running
{
  while [ "1" ]; do
        sleep 1
  done
}
function getConfigLtsKey
{
        local ltsconffile="$1"
        local varname="$2"
        . $ltsconffile

        eval "echo \$$varname"
}
```

Create executable service start/stop file **/usr/share/ltsp/client/custom/vconsole**
```
#!/bin/bash

action="$1"

. $(dirname $0)/common

interval=2

function openvt_xfreerdp
{
        local i=$i
        if [ "$i" -ne "7" ] ; then
                openvt -f -c$i --  su -c "xinit $customscriptdir/vconsole-xfreerdp $i -display :$i -- :$i vt$i"
        fi
}

exit_script() {
        trap - SIGINT SIGTERM # clear the trap
        do_stop
        kill -- -$$ # Sends SIGTERM to child/sub processes
}

function do_start
{
#Ensure all vconsoles are started to avoid agetty starting after us. 
#Starting it without messing with agetty taking over consoles 2-6 maybe should be done by correctly either disabling it or ensure it has finished starting with dependencies inside the systemd configuration file.
#For now a workaround is used.
        for i in 2 3 4 5 6; do
                chvt $i
                sleep 1
        done

        for i in {2..15} ; do
                if [ "${#i}" = "1" ]; then
                        screen=0$i;
                else
                        screen=$i;
                fi
                cmdstr=$(getConfigLtsKey /etc/ltsp/lts.conf SCREEN_$screen)

                if [ "$cmdstr" ]; then
                        sleep $interval

                        openvt_xfreerdp $i &
                fi
        done
        sleep $interval
        chvt 7
        keep_running
}

function do_stop
{
        killall vconsole-xfreerdp &> /dev/null
        exit 0
}

trap exit_script SIGINT SIGTERM

case $action in
        start) do_start ;;
        stop) do_stop ;;
esac
```
``chmod +x /usr/share/ltsp/client/custom/vconsole``

Create file **/usr/share/ltsp/client/custom/vconsole-xfreerdp**
```
#!/bin/bash

. $(dirname $0)/common

screen="$1"

if [ "${#screen}" = "1" ]; then
        screen=0$screen;
fi

cmdstr=$(getConfigLtsKey /etc/ltsp/lts.conf SCREEN_$screen)

[ "$cmdstr" ] || exit 0
while [ "1" ]; do
        $cmdstr
done
```
``chmod +x /usr/share/ltsp/client/custom/vconsole-xfreerdp``

Create a conf file /etc/ltsp/lts.conf similar to old lts.conf 
```
SCREEN_06="xfreerdp /f /compression /bpp:16  /cert-ignore -sec-nla /v:192.168.1.15"
SCREEN_09="xfreerdp /f /compression /bpp:16  /cert-ignore -sec-nla /v:192.168.1.15"
```

Before running ltsp image / execute
``systemctl enable ltsp-vconsole.service``
After creating ltsp image we can disable the service
``systemctl disable ltsp-vconsole.service``

Is there a place i can put these commands (like the folder for scripts for initrd /usr/share/ltsp/common/initrd/) to run them before and after creating the image file?

**Comment by alkisg** [comment]
Eeeh that's way too complicated and too very little integrated! Maybe a starting point would be just this:

```shell
[clients]
POST_SERVICE_SCREEN_08="openvt -fc8 startx -- xterm"
```

Inside that xterm, test the command that you want, e.g. xfreerdp. Then just put the command inside ltsp.conf....

**Comment by icevk77** [comment]
The command 
``openvt -fc8 startx -- xterm``
worked but not on every rerun. In most reboots it didn't. It even dropped me on virtual console 1, instead of graphic login.
```
Fatal server error. 
(EE)Service is already active for display :0
```
The same happened with 
``openvt -fc8 startx -- /usr/bin/xfreerdp /f /cert-ignore -sec-nla /v:192.168.1.15``

I tried also xinit. LTSP5 used xinit as well. For example.
``openvt -f -c8 -- xinit /usr/bin/xfreerdp /f /cert-ignore -sec-nla /v:192.168.1.15 -display :8 -- :8 vt8``

However this comand did not work when put inside ltsp.conf. 

I logged in as root to the ltsp client and tried the above command and various others that worked. 

Unfortunately by putting them in the ltsp.conf file, they did not work on client boot.

**Comment by alkisg** [comment]
It works for me in every boot. It also works with xinit, for example:

```shell
POST_SERVICE_SCREEN_08="sleep 10; openvt -fc8 -- xinit /usr/bin/xterm -display :8 -- :8 vt8"
```

For me, it works with or without `sleep 10` there. Is it a race condition for you, does it work if you add that `sleep`?

In any case, a proper implementation should work for both xorg and wayland, and it should disable the getty service at vt8 to prohibit it from interfering. But as I said, sorry but at this time I don't have enough free time to research this; I can propose some ideas if I have any, but that's about it. :)

**Reply by icevk77** [reply]
No it didn't work with a sleep 5, i had already tested that.
I will try to increment this to 10 or more to see what happens.

The getty service is indeed a problem to virtual consoles 2-6 where it automatically starts and can take over if it starts after our command. 
For virtual consoles 8 and above it will not start so it is not a problem there.

Especially for console 6 it may start even after our command as it is kind of a reserved defined virtual console.

To disable getty from using virtual consoles 2-6 and also disable the reserved virtual console 6, a change in file **/etc/systemd/logind.conf** is needed

```
NAutoVTs=1
ReserveVT=0
```
I will try more and if i can find a way I will post it here

Thank you for your help

**Comment by alkisg** [comment]
This also works fine for me on vt6, it's the best so far:

```shell
POST_INIT_SCREEN_06="mkdir -p /etc/systemd/system/getty@tty6.service.d/
echo '[Service]
ExecStart=
ExecStart=-/usr/bin/xinit /usr/bin/xterm -display :6 -- :6 vt6' >/etc/systemd/system/getty@tty6.service.d/override.conf"
```

It's based on: https://wiki.archlinux.org/title/getty#Automatic_login_to_virtual_console

**Comment by icevk77** [comment]
Interesting way and works consistently on every reboot for xterm.

I am trying to get it to work with xfreerdp and I will post my findings when i manage it.
Thanks.

**Comment by alkisg** [comment]
Linking to a similar question/answer: https://github.com/ltsp/ltsp/discussions/205
